docs/design/garbage-collection.md

```diff
@@ -145,3 +146,74 @@ see: [Bloom filter math](http://pages.cs.wisc.edu/~cao/papers/summary-cache/node
+- **Custom**: Custom bloom filter
+
+### Zeebo's bloom filters
+- Parameters:
+    - **k**: The Bloom filter will be built such that the probability of a false positive is less than (1/2)^k
+    - **h**: hash functions
+- Serialization available
+- hash functions are to be given as a parameter to the constructor
+### Willf's bloom filters
+- Parameters:
+    - **m**: max size in bits
+    - **k**: number of hash functions
+- hash functions not configurable
+
+### Steakknife's bloom filters
+- Parameters:
+    - **maxElements**: max number of elements in the set
+    - **p**: probability of false positive
+- Serialization available
+- murmur3 hash function
+
+### Custom bloom filter
+- Parameters:
+    - **maxElements**: max number of elements in the set
+    - **p**: probability of false positive
+- The piece id is used as a hash function.
+
+
+### Benchmark
+We assume a typical storage nodes has 2 TB capacity, and a typical piece is ~2 MB, so we are testing the behavior with 1 million pieces.
+
+We create a list of 1 million piece ids and add 95% of them to the Bloom filter. We then check if the 95% are contained in the set (there should be no false negative) and we evaluate the false positive rate by checking the remaining 5% piece ids.
+
+For each target false positive probability between 1% and 20% and each bloom filter type, we measure the size (in bytes) of the encoded bloom filter and the observed false positive rate.
+
+
+|p|	Zeebo size|Zeebo real_p| Willf size|Willf real_p|Steakknife size|Steakknife real_p| Custom size|Custom real_p|
+|---|	---|	---|	---|	---|	---|	---|	---|	---|
+|0.01|	9437456|	0.01|	1198160|	0.01|	1198264|	0.01|	1250012|	0.01|
+|0.02|	8913247|	0.02|	1017824|	0.02|	1017920|	0.02|	1125012|	0.01|
+|0.03|	8913247|	0.02|	912336|	    0.03|	912432|	0.03|	1000012|	0.02	|
+|0.04|	8389038|	0.03|	837488|	    0.04|	837576|	0.03|	875012|	0.03|
+|0.05|	8389038|	0.03|	779432|	    0.04|	779520|	0.04|	875012|	0.03|
+|0.06|	8389038|	0.03|	732000|	    0.06|	732088|	0.05|	750012|	0.05|
+|0.07|	7864829|	0.06|	691888|	    0.06|	691968|	0.06|	750012|	0.05|
+|0.08|	7864829|	0.06|	657152|	    0.07|	657232|	0.07|	750012|	0.05|
+|0.09|	7864829|	0.06|	626504|	    0.08|	626584|	0.08|	750012|	0.05|
+|0.10|	7864829|	0.06|	599096|	    0.09|	599176|	0.09|	625012|	0.08|
+|0.11|	7864829|	0.06|	574296|	    0.10|	574376|	0.10|	625012|	0.08|
+|0.12|	7864829|	0.06|	551656|	    0.11|	551736|	0.11|	625012|	0.08|
+|0.13|	7340620|	0.12|	530832|	    0.11|	530904|	0.12|	625012|	0.08|
+|0.14|	7340620|	0.12|	511552|	    0.12|	511624|	0.13|	625012|	0.08|
+|0.15|	7340620|	0.12|	493600|	    0.14|	493672|	0.14|	500012|	0.16|
+|0.16|	7340620|	0.12|	476816|	    0.15|	476888|	0.15|	500012|	0.16|
+|0.17|	7340620|	0.12|	461040|	    0.15|	461112|	0.16|	500012|	0.16|
+|0.18|	7340620|	0.12|	446168|	    0.17|	446240|	0.17|	500012|	0.16|
+|0.19|	7340620|    0.12|	432104|	    0.18|	432176|	0.18|	500012|	0.16|
+|0.20|	7340620|	0.12|	418760|	    0.19|	418832|	0.19|	500012|	0.16|
```

Comment:
So one disadvantage of the scalable bloom filter is that it may allocate a new, 2x as large filter, right before you're done adding values. Sort of like having a byte slice of size 4096, appending one value, and ending up with one of capacity 8192.

I do expect the sbloom implementation to be larger, but not 10x larger. I wonder if it's very sparse and would compress well. I don't expect the other filters to compress well if they're tuned correctly: at optimal tuning, a bloom filter is 50% 1 bits randomly distributed. 

Can we try a quick gzip around the sbloom output and see if that makes a big difference?